### PR TITLE
Fixes for ROS2 tests

### PR DESCRIPTION
**Pull request by Tadinu**
`RRROS2GameMode` 
* Fix Sim init time measurement add `BeginTimestampInSec`
* Add `NativeSpawnableClasses`, being configured directly in GameMode ctor instead of GameState, also auto-registering them
`ATurtlebotBurger` fix root comp issue which affect `GetActorTransform()`

`URRObjectUtils::CreateMeshComponent()` fix `bStationary` in init mesh comp

External ROS2 tests
* robot_spawn/remove -> entity_spawn/remove
* utils: add `wait_for_entity_moving()`, used by `test_robot_cmd_vel_subscribed`. Move `CmdVelPublisher` to `cmd_vel_publisher.py`
* `wait_for_spawned_entity()` add `in_entity_ref_frame` param
* Add `BasePublisher, BaseSubscriber` classes, inherited by CmdVel, Path publisher, ModelState, TF subscriber

### CODE REVIEW

File: Source/RapyutaSimulationPlugins/Private/Core/RRROS2GameMode.cpp [hunk 1]
```diff
@@ -11,12 +11,25 @@
...
+#include "Robots/Turtlebot3/TurtlebotBurger.h"
+#include "Robots/Turtlebot3/TurtlebotBurgerVehicle.h"
 #include "Tools/RRGhostPlayerPawn.h"
 #include "Tools/RRROS2ClockPublisher.h"
 
 ARRROS2GameMode::ARRROS2GameMode()
 {
     DefaultPawnClass = ARRGhostPlayerPawn::StaticClass();
+
+    // Static spawnable classes
```
Comment (yuokamoto) [inline Source/RapyutaSimulationPlugins/Private/Core/RRROS2GameMode.cpp]: should avoid adding thses static classes in this base class for test.

File: Source/RapyutaSimulationPlugins/Private/Robots/Turtlebot3/TurtlebotBurger.cpp [hunk 2]
```diff
@@ -55,6 +59,17 @@ void ATurtlebotBurger::SetupBody()
     SetupConstraintsAndPhysics();
 }
 
+void ATurtlebotBurger::PreInitializeComponents()
+{
+    Super::PreInitializeComponents();
+    if (IsDynamicRuntimeRobot() && DefaultRoot->GetAttachParent())
+    {
+        // NOTE: This requires [DefaultRoot] to have been attached to some NEW Root, by which there is no need to promote its child
```
Comment (yuokamoto) [inline Source/RapyutaSimulationPlugins/Private/Robots/Turtlebot3/TurtlebotBurger.cpp]: when "[DefaultRoot] to have been attached to some NEW Root" happen?

File: Source/RapyutaSimulationPlugins/Private/Robots/Turtlebot3/TurtlebotBurger.cpp [hunk 3]
```diff
@@ -27,9 +27,13 @@ void ATurtlebotBurger::SetupBody()
     WheelRight = CreateDefaultSubobject<UStaticMeshComponent>(TEXT("WheelRight"));
     CasterBack = CreateDefaultSubobject<UStaticMeshComponent>(TEXT("CasterBack"));
 
-    if (RootComponent)
+    // NOTE: This is critical, otherwise GetTransform() always returns one of [DefaultRoot] as initial one at spawning due to its being just a non-physics scene component
```
Comment (yuokamoto) [inline Source/RapyutaSimulationPlugins/Private/Robots/Turtlebot3/TurtlebotBurger.cpp]: - Do you mean this is a issue at the spawning time only?
- If this is a issue to general robot, shouldn't be part of base robot?

File: Source/RapyutaSimulationPlugins/Private/Core/RRROS2GameMode.cpp (hunk 1 above)
Comment (Tadinu) [reply Source/RapyutaSimulationPlugins/Private/Core/RRROS2GameMode.cpp]: So basically, this is the place configuring spawnable entity classes, which is not only for test, but more to accommodate `SpawnEntity` srv. Otherwise, they have to be manually configured in all maps' Level Blueprint.

File: Source/RapyutaSimulationPlugins/Private/Core/RRROS2GameMode.cpp [hunk 4]
```diff
@@ -113,6 +127,7 @@ void ARRROS2GameMode::SetFixedTimeStep(const float InStepSize)
     }
     FApp::SetUseFixedTimeStep(true);
     FApp::SetFixedDeltaTime(InStepSize);
+    UE_LOG_WITH_INFO_SHORT(LogRapyutaCore, Display, TEXT("FIXED TIMESTEP UPDATED: %f"), InStepSize);
```
Comment (yuokamoto) [inline Source/RapyutaSimulationPlugins/Private/Core/RRROS2GameMode.cpp]: Better to use Capital only beginning of word for consistency

File: Source/RapyutaSimulationPlugins/Private/Robots/Turtlebot3/TurtlebotBurger.cpp (hunk 2 above)
Comment (Tadinu) [reply Source/RapyutaSimulationPlugins/Private/Robots/Turtlebot3/TurtlebotBurger.cpp]: That only happens if the RootComp has been switched to a new one earlier, and thus parenting DefaultRoot. But this part is actually unrequired here due to TurtlebotBurger is statically-built class, so removed.

File: Source/RapyutaSimulationPlugins/Private/Robots/Turtlebot3/TurtlebotBurger.cpp (hunk 3 above)
Comment (Tadinu) [reply Source/RapyutaSimulationPlugins/Private/Robots/Turtlebot3/TurtlebotBurger.cpp]: Yeah I added `SetBaseMeshComp()` for RRBaseRobot, so child classes, if creating a custom base mesh comp, could use to automatically make it new RootComp.

File: Source/RapyutaSimulationPlugins/Private/Core/RRROS2GameMode.cpp (hunk 1 above)
Comment (yuokamoto) [reply Source/RapyutaSimulationPlugins/Private/Core/RRROS2GameMode.cpp]: move to .ini

File: Source/RapyutaSimulationPlugins/Private/Robots/RRBaseRobot.cpp [hunk 5]
```diff
@@ -185,6 +185,32 @@ void ARRBaseRobot::PostInitializeComponents()
     }
 }
 
+void ARRBaseRobot::SetBaseMeshComp(UMeshComponent* InBaseMeshComp, bool bInMakeAsRoot, bool bInDestroyDefaultRoot)
+{
+    ensure(InBaseMeshComp);
```
Comment (yuokamoto) [inline Source/RapyutaSimulationPlugins/Private/Robots/RRBaseRobot.cpp]: should return if InBaseMeshComp == nullptr?

File: Content/SkeletalRobots/Turtlebot3/BallCasterSphereWheeled/BP_BallCasterSphereWheeledTurtlebotBurger.uasset
Comment (yuokamoto) [inline Content/SkeletalRobots/Turtlebot3/BallCasterSphereWheeled/BP_BallCasterSphereWheeledTurtlebotBurger.uasset:1]: is this just renamed? 
is this related to the test?

File: ExternalTest/rr_sim_tests/rr_sim_tests/utils/path_publisher.py [hunk 6]
```diff
@@ -18,70 +18,39 @@
...
+PATH_PUBLISH_FREQ = 10.0 #Hz
+class PathPublisher(BasePublisher):
+    def __init__(self, in_robot_namespace:str = '',
+                 in_topic_name = TOPIC_NAME_PATH,
+                 in_publish_num = PATH_PUBLISH_NUM,
+                 in_publish_freq = PATH_PUBLISH_FREQ,
+                 in_robot_path:Path = Path()):
+        super().__init__(in_robot_namespace, in_topic_name, in_publish_num, in_publish_freq)
+        self._publisher = self.create_publisher(
+            Path, self._full_topic_name, qos_profile=PUBLISHER_DEFAULT_QOS
```
Comment (yuokamoto) [inline ExternalTest/rr_sim_tests/rr_sim_tests/utils/path_publisher.py]: better to use in_topic_name directly. 
Name resolution should be done by ROS

File: ExternalTest/rr_sim_tests/rr_sim_tests/utils/base_subscriber.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    def __init__(self, in_robot_namespace: str, 
+                       in_topic_name: str,
+                       in_subscriber_rate=SUBSCRIBER_DEFAULT_RATE):
+        super().__init__(
+            node_name=f"{in_topic_name.replace('/','_')}_subscriber", namespace=in_robot_namespace
+        )
+
+        # https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/node_impl.hpp
+        # https://github.com/ros2/rclcpp/issues/1767
+        self._full_topic_name = None
```
Comment (yuokamoto) [inline ExternalTest/rr_sim_tests/rr_sim_tests/utils/base_subscriber.py]: if this is a common feature for both pub and sub, should be separated

File: ExternalTest/rr_sim_tests/rr_sim_tests/utils/base_publisher.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        in_publish_num=PUBLISHER_DEFAULT_PUB_NUM,
+        in_publish_freq=PUBLISHER_DEFAULT_FREQ,
+        in_auto_publish=True
+    ):
+        super().__init__(node_name=f"{in_topic_name.replace('/','_')}_publisher", namespace=in_robot_namespace)
+
+        # https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/node_impl.hpp
+        # https://github.com/ros2/rclcpp/issues/1767
+        self._full_topic_name = None
+        if in_topic_name.startswith('/') or in_topic_name.startswith('~'):
```
Comment (yuokamoto) [inline ExternalTest/rr_sim_tests/rr_sim_tests/utils/base_publisher.py]: ~ means private name space, i.e. it become nodename/topic name

Comment (yuokamoto) [reply ExternalTest/rr_sim_tests/rr_sim_tests/utils/base_publisher.py]: Does get_namespace include node name as well?
topic name should become ns/node_name/topic_name

File: Content/SkeletalRobots/Turtlebot3/BallCasterSphereWheeled/BP_BallCasterSphereWheeledTurtlebotBurger.uasset
Comment (Tadinu) [reply Content/SkeletalRobots/Turtlebot3/BallCasterSphereWheeled/BP_BallCasterSphereWheeledTurtlebotBurger.uasset:1]: yes, there is a bug related to searching for `BP_TurtlebotBurger` as defined in `BPSpawnableClassNames`, each of which must hold a unique name, so the spawning could find the correct class. They should always be regardless anw.

File: ExternalTest/rr_sim_tests/rr_sim_tests/utils/base_publisher.py (hunk 8 above)
Comment (Tadinu) [reply ExternalTest/rr_sim_tests/rr_sim_tests/utils/base_publisher.py]: So it should be `get_effective_namespace()` in case of ~ I guess

File: ExternalTest/rr_sim_tests/rr_sim_tests/utils/base_subscriber.py (hunk 7 above)
Comment (Tadinu) [reply ExternalTest/rr_sim_tests/rr_sim_tests/utils/base_subscriber.py]: You meant grouped?, I so added `base_pub_sub_node.py`

**Review by yuokamoto — APPROVED** [review]
@chalice-graeme please review
